### ISSUE TITLE
`from_pretrained` with modified `num_channels` produces inf/nan outputs

### ISSUE THREAD

**Issue by t-muser** [camlab-ethz/poseidon#18]
I'm trying to fine-tune the `tiny` model but I keep producing extremely large predictions which result in `nan` gradients, making it impossible to train the model. 

```python
import torch
# from scOT.model import ScOT 

def make_POSEIDON(dim_in, dim_out, spatial_resolution, from_pretrained="camlab-ethz/Poseidon-T"):
    image_size = spatial_resolution[0]
    if not all([res_i == image_size for res_i in spatial_resolution]):
        raise ValueError
    
    # seems like the easiest way to override dim_in/dim_out while keeping everything else the same
    model = ScOT.from_pretrained(
        from_pretrained
    )
    config = model.config
    config.image_size = image_size
    config.num_channels = dim_in
    config.num_out_channels = dim_out
    model = ScOT.from_pretrained(
        from_pretrained,
        config=config,
        ignore_mismatched_sizes=True
    )
    return model

dim_in = 5 # anything except 4 (which retains the pre-trained model's layers) causes output to explode
model = make_POSEIDON(dim_in=dim_in, dim_out=dim_in, spatial_resolution=(128,128))
x = torch.rand((45, dim_in, 128, 128))
model(x, time=0.1 * torch.ones(x.shape[0])).output # will be around 1e27 to inf/nan
```

Environment:
``` 
requires-python = "~=3.13.0"
dependencies = [
    "torch >= 2.0.1",
    "torchvision >= 0.15.2",
    "numpy < 2.0.0",
    "transformers == 4.55.0",
    "matplotlib",
    "accelerate>=0.32.0",
    "wandb==0.22.2",
    "h5py",
    "pandas",
    "pyyaml",
    "netcdf4>=1.7.2",
    "einops>=0.8.1",
    "scipy>=1.16.1",
    "pytorch-lightning>=2.3.3",
    "ninja>=1.13.0",
    "ipykernel>=6.30.1",
    "seaborn>=0.13.2",
    "huggingface-hub[cli]>=0.34.4",
    "xarray>=2025.8.0",
    "torchinfo>=1.8.0",
    "the-well>=1.1.0",
    "hydra-core>=1.3.2",
    "denoising-diffusion-pytorch>=2.2.5",
    "timm>=1.0.20",
    "neuraloperator>=2.0.0",
    "triton>=3.4.0",
    "ruff>=0.14.5",
]
```

**Comment by lamthuy** [comment]
I also got the same issue, after probing the code I see that when num channels differs from the pretrained model's num channels the embeddings and patch recovery layers have some mismatching sizes, it initialise the weights of those layers with values that might not be in safe zone of the pretrained models so causing inf and nan. I think the following code seems to wrk:

```
import torch


# from scOT.model import ScOT
from scOT.model import ScOT


def make_POSEIDON(dim_in, dim_out, spatial_resolution, from_pretrained="camlab-ethz/Poseidon-T"):
    image_size = spatial_resolution[0]
    if not all([res_i == image_size for res_i in spatial_resolution]):
        raise ValueError

    # seems like the easiest way to override dim_in/dim_out while keeping everything else the same
    model = ScOT.from_pretrained(
        from_pretrained
    )
    config = model.config
    config.image_size = image_size
    config.num_channels = dim_in
    config.num_out_channels = dim_out
    model = ScOT.from_pretrained(
        from_pretrained,
        config=config,
        ignore_mismatched_sizes=True
    )
    for n, p in model.embeddings.patch_embeddings.named_parameters():
        torch.nn.init.normal_(p, mean=0.0, std=0.01)  # smaller std than default

    for n, p in model.patch_recovery.named_parameters():
        torch.nn.init.normal_(p, mean=0.0, std=0.01)  # smaller std than default
    return model


dim_in = 5  # anything except 4 (which retains the pre-trained model's layers) causes output to explode
model = make_POSEIDON(dim_in=dim_in, dim_out=dim_in, spatial_resolution=(128, 128))
x = torch.rand((45, dim_in, 128, 128))
print(model(x, time=0.1 * torch.ones(x.shape[0])).output)
```

**Comment by t-muser** [comment]
Hey @lamthuy,

thanks for the help! This does indeed seem to do the trick and give reasonable values.

I would like to keep this issue open just to see if the authors consider this the recommended way to deal with the problem.

**Comment by maximilianherde** [comment]
That is a problem with the HF Transformers version that I encountered myself at some point. When we ran the experiments it was not a problem but later they made some changes in the library which results in this problem. I might need to push a fix for that....
But yes, in principle the above solution is correct, you need to init the weights yourself. For the correct HF version, you would need to provide a `_init_weights` function in the model class.
